sys/arm64/vmm/vmm_hyp.c

```diff
@@ -0,0 +1,800 @@
+	 * invalidate the stage 1 TLB. As we don't know which stage 1 virtual
+	 * addresses point at the stage 2 IPA we need to invalidate the entire
+	 * stage 1 TLB.
+	 */
+
+	start = TLBI_VA(sva);
+	end = TLBI_VA(eva);
+	for (r = start; r < end; r += TLBI_VA_L3_INCR) {
+		/* Invalidate the stage 2 TLB entry */
+		if (final_only)
+			__asm __volatile("tlbi	ipas2le1is, %0" : : "r"(r));
+		else
+			__asm __volatile("tlbi	ipas2e1is, %0" : : "r"(r));
+	}
+	/* Ensure the entry has been invalidated */
+	dsb(ish);
+	/* Invalidate the stage 1 TLB. */
+	__asm __volatile("tlbi vmalle1is");
+	dsb(ish);
+	isb();
+
+	/* Switch back t othe host vttbr */
+	WRITE_SPECIALREG(vttbr_el2, host_vttbr);
+	isb();
+
+	return (0);
+}
+
+static int
+vm_s2_tlbi_all(uint64_t vttbr)
+{
+	uint64_t host_vttbr;
+
+	/* Switch to the guest vttbr */
+	/* TODO: Handle Cortex-A57/A72 erratum 131936 */
+	host_vttbr = READ_SPECIALREG(vttbr_el2);
+	WRITE_SPECIALREG(vttbr_el2, vttbr);
+	isb();
+
+	__asm __volatile("tlbi vmalls12e1is");
+	dsb(ish);
+	isb();
+
+	/* Switch back t othe host vttbr */
+	WRITE_SPECIALREG(vttbr_el2, host_vttbr);
+	isb();
+
+	return (0);
+}
+
+static int
+vmm_dc_civac(uintptr_t start, uint64_t len)
+{
+	size_t line_size, end;
+	uint64_t ctr;
+
+	ctr = READ_SPECIALREG(ctr_el0);
+	line_size = sizeof(int) << CTR_DLINE_SIZE(ctr);
+	end = start + len;
+	dsb(ishst);
+	/* Clean and Invalidate the D-cache */
+	for (; start < end; start += line_size)
+#ifdef __CHERI_PURE_CAPABILITY__
+		__asm __volatile("dc	civac, %0" :: "C" (start) : "memory");
+#else
+		__asm __volatile("dc	civac, %0" :: "r" (start) : "memory");
+#endif
+	dsb(ish);
+	return (0);
+}
+
+static int
+vmm_el2_tlbi(uint64_t type, uint64_t start, uint64_t len)
+{
+	uint64_t end, r;
+
+	dsb(ishst);
+	switch (type) {
+	default:
+	case HYP_EL2_TLBI_ALL:
+		__asm __volatile("tlbi	alle2" ::: "memory");
+		break;
+	case HYP_EL2_TLBI_VA:
+		end = (start + len) >> 12;
+		start >>= 12;
+		while (start < end) {
+			/* TODO: Use new macros when merged past them */
+			r = start & 0xffffffffffful;
+			__asm __volatile("tlbi	vae2is, %0" :: "r"(r));
+			start += PAGE_SIZE;
+		}
+		break;
+	}
+	dsb(ish);
+
+	return (0);
+}
+
+uint64_t
+vmm_hyp_enter(uint64_t handle, uintptr_t x1, uintptr_t x2, uintptr_t x3,
+    uintptr_t x4, uintptr_t x5, uintptr_t x6, uintptr_t x7)
+{
+	uint64_t ret;
+
+	switch (handle) {
+	case HYP_ENTER_GUEST:
+		do {
+			ret = vmm_hyp_call_guest((struct hyp *)x1,
+			    (struct hypctx *)x2);
+		} while (ret == EXCP_TYPE_REENTER);
+		return (ret);
+	case HYP_READ_REGISTER:
+		return (vmm_hyp_read_reg(x1));
+	case HYP_CLEAN_S2_TLBI:
+		return (vmm_clean_s2_tlbi());
+	case HYP_DC_CIVAC:
+		return (vmm_dc_civac(x1, x2));
+	case HYP_EL2_TLBI:
+		return (vmm_el2_tlbi(x1, x2, x3));
+	case HYP_S2_TLBI_RANGE:
+		return (vm_s2_tlbi_range(x1, x2, x3, x4));
+	case HYP_S2_TLBI_ALL:
+		return (vm_s2_tlbi_all(x1));
+	case HYP_CLEANUP:	/* Handled in vmm_hyp_exception.S */
+	default:
```

Comment:
Should this not be an error?

Reply:
It could be (return 1 perhaps?), but most callers don't check for errors anyway, and the meaning of the return value depends on the verb passed in. It's not really nice to panic here since this code doesn't run in the kernel's usual context.